vulture_os/toolkit/api_parser/cybereason/cybereason.py

```diff
@@ -517,37 +527,37 @@ def format_log(self, kind, log):
         return json.dumps(log)
 
     def execute(self):
-        # Get last api call or 30 days ago
-        since = self.frontend.last_api_call or (timezone.now() - timedelta(days=30))
-        # 24h max per request
-        to = min(timezone.now(), since + timedelta(hours=24))
-
-        # delay the times of 5 minutes, to let the times at the API to have all logs
-        to = to - timedelta(minutes=5)
+        for kind in ["malops", "malwares"]:
+            # Get last api call or 30 days ago
+            since = self.last_collected_timestamps.get(kind, self.frontend.last_api_call) or (timezone.now() - timedelta(days=30))
+            # 24h max per request + Don't get the last 5 minutes of logs, as some can appear delayed by the API
+            to = min(timezone.now() - timedelta(minutes=5), since + timedelta(hours=24))
 
-        msg = f"Parser starting from {since} to {to}"
-        logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
+            msg = f"{kind}'s parser starting from {since} to {to}"
+            logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
 
-        for kind in ["malops", "malwares"]:
             logs = self.get_logs(kind, since, to)
-
             # Downloading may take some while, so refresh token in Redis
             self.update_lock()
 
-            enriched_logs = [
-                self.format_log(kind, self.add_enrichment(kind, log))
-                for log in logs
-            ]
-
+            enriched_logs = [self.format_log(kind, self.add_enrichment(kind, log)) for log in logs]
             # Enriching may take some while, so refresh token in Redis
             self.update_lock()
 
             self.write_to_file(enriched_logs)
             # Writting may take some while, so refresh token in Redis
             self.update_lock()
 
-        # update last_api_call only if logs are retrieved
-        self.frontend.last_api_call = to
+            if len(logs) >= 10000: # ELS don't supports sending more than 10k results by query -- maybe this never happens due to current implementation of pagination, by security I prefer adding this check
+                msg = f"{kind}: retrieved {len(logs)} lines >= 10k logs, shortening time range to avoid API error"
+                logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
+                to -= (to - since)/2
+                logs = list()
+                continue
+
+            if logs or (to - since) >= timedelta(hours=23, minutes=55): # If logs or no logs during last 24h we want to update the timestamp
+                self.last_collected_timestamps[kind] = to
+                self.frontend.save()
```

Comment:
saving the frontend is useless here, it will always be saved at the end of a run, even in case of error
```suggestion
            if logs or (to - since) >= timedelta(hours=23, minutes=55): # If logs or no logs during last 24h we want to update the timestamp
                self.last_collected_timestamps[kind] = to
```